tests/integration/test_charm.py

```diff
@@ -37,24 +42,232 @@ async def test_build_and_deploy(ops_test: OpsTest):
+    random_unit = ops_test.model.applications[APP_NAME].units[0]
+    server_config_credentials = await get_server_config_credentials(random_unit)
+
+    primary_unit = await get_primary_unit(
+        ops_test,
+        random_unit,
+        server_config_credentials["username"],
+        server_config_credentials["password"],
+    )
+
+    random_chars = generate_random_string(40)
+    create_records_sql = (
+        "CREATE DATABASE IF NOT EXISTS test",
+        "CREATE TABLE IF NOT EXISTS test.data_replication_table (id varchar(40), primary key(id))",
+        f"INSERT INTO test.data_replication_table VALUES ('{random_chars}')",
+    )
+
+    mysqlsh_sql_base_command = f"mysqlsh --sql {server_config_credentials['username']}:{server_config_credentials['password']}@127.0.0.1"
+
+    await run_command_on_unit(
+        primary_unit, f"{mysqlsh_sql_base_command} -e \"{';'.join(create_records_sql)}\""
+    )
+
+    # Allow time for the data to be replicated
+    time.sleep(2)
+
+    # Confirm that the values are available on all units
     for unit in ops_test.model.applications[APP_NAME].units:
-        # Ensure that mysql-server is installed correctly
-        result = await run_command_on_unit(unit, "mysqld --version")
-        mysql_version = result.strip().split()
-        assert mysql_version[2] == "8.0.28-0ubuntu0.20.04.3"
-
-        # Ensure that mysql-shell is installed correctly
-        result = await run_command_on_unit(unit, "mysqlsh --version")
-        mysqlsh_version = result.strip().split()
-        assert mysqlsh_version[2] == "8.0.23"
+        output = await run_command_on_unit(
+            unit,
+            f"{mysqlsh_sql_base_command} -e \"SELECT * FROM test.data_replication_table WHERE id = '{random_chars}'\"",
+        )
+        assert random_chars in output
+
+
+@pytest.mark.order(3)
+@pytest.mark.abort_on_fail
+async def test_primary_reelection(ops_test: OpsTest):
+    """Confirm that a new primary is elected when the current primary is torn down."""
+    random_unit = ops_test.model.applications[APP_NAME].units[0]
+    server_config_credentials = await get_server_config_credentials(random_unit)
+
+    primary_unit = await get_primary_unit(
+        ops_test,
+        random_unit,
+        server_config_credentials["username"],
+        server_config_credentials["password"],
+    )
+    primary_unit_name = primary_unit.name
+
+    assert len(ops_test.model.applications[APP_NAME].units) == 3
+
+    # Destroy the primary unit and wait 5 seconds to ensure that the
+    # juju status changed from active
+    await ops_test.model.destroy_units(primary_unit.name)
+    time.sleep(5)
+    await ops_test.model.wait_for_idle(
+        apps=[APP_NAME],
+        status="active",
+        raise_on_blocked=True,
+        timeout=1000,
+    )
+
+    # Wait for unit to be destroyed and confirm that the new primary unit is different
+    assert len(ops_test.model.applications[APP_NAME].units) == 2
```

Comment:
Though I like the `@pytest.mark.order(x)` flags, you could make the order more flexible by scaling back up to 3 units at the end of this test.

Reply:
That's a great suggestion! It's good to assume idempotency event if there's an established order of execution for tests. Addressed in [7972d12](https://github.com/canonical/mysql-operator/pull/10/commits/7972d128c39bbf46d38369d4579cc2808d34c85d)